src/ansi-c/c_typecheck_code.cpp

```diff
@@ -809,6 +818,41 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
 
   typecheck_spec_loop_invariant(code);
   typecheck_spec_decreases(code);
+
+  if(code.find(ID_C_spec_assigns).is_not_nil())
+    typecheck_spec_assigns(
+      static_cast<unary_exprt &>(code.add(ID_C_spec_assigns)).op().operands());
+}
```

Comment:
Perhaps we need an example with `do { ... } while` to cover this case?

Reply:
Yes, which I think requires other fixes to work correctly as well as we discussed this morning. I would leave it as future work for now and raise an exception on contract application on `do ... while` loops.